proposals/0015-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,355 @@
+reward distribution phase.
+
+
+```
+struct EpochRewardReserve {
+   reserve_hash: Hash,
+   reward_balance: u64,
+}
+```
+
+The account balance of `RewardsReserve` is equal to the total amount (in the
+`AccountInfo::lamports` field) of all rewards to be distributed in a block. And
+the `reward_balance` field shadow the `AccountInfo::lamports`. The
+`reserve_hash` is computed from the rewards in the block as follows.
+
+```
+reserve_hash = hash([(stake_account_key, reward_amount)])
+```
+
+At the end of reward computation phase, `M` `EpochRewardsReserve` sysvar
+accounts are updated, where `M` corresponds to how many blocks will be in
+the reward distribution phase, which is described in the next section.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks. Each of the `M` blocks is responsible for distributing
+the reward from one partition of the rewards stored in the `EpochRewardReserve`
+account with the address specified by the `get_epoch_reward_reserve_addr`
+function above. Once all rewards have been distributed, the balance of the
+`EpochRewardReserve` account and shadow balance field `reward_balance` both
+should be reduced to `0`. However, to protect from unexpected deposits to
+`EpochRewardReserve` accounts, when the reward distribution completes, any
+extra lamports in `EpochRewardReserve` accounts will be burned.
+
+Before each reward distribution, the `EpochRewardReserve` account's
+`reserve_balance` is checked to make sure there is enough balance to distribute
+the reward. After a reward is distributed, the balance and `reserve_balance` in
+`EpochRewardReserve` account is decreased by the amount of the reward that was
+distributed. The hash of reward account address and reward amount is
+accumulated when rewards are distributed. After all the rewards are distributed
+from the reserve account for the block, the hash computed from all the rewards
+distribution is verified against the hash stored in the `EpochRewardReserve`
+account to make sure they match.
+
+During the reward distribution phase, there is a "root accumulator hash" that
+starts empty and is gradually filled by adding each hash of a
+`EpochRewardReserve`. The distribution process continues until the "root
+accumulator hash" matches the "root hash" stored in `EpochRewardHistory`. At this
+point, the reward distribution for the current epoch is complete and we can
+safely assume that there will be no further distributions for the remainder of
+the epoch.
+
+### RPC API Changes
+
+To accommodate the new reward distribution approach, there are a few changes
+to the rewards related RPC APIs.
+
+In particular, `getInflationReward` RPC API will change. It will be unavailable
+during reward period, and only available after reward distribution completes.
+
+To help to determine when the reward period completes, a new RPC API,
+`getRewardPeriodInfo`,  will be added. This API will returns:
```

Comment:
this info should probably be made available via a sysvar/syscall as well, so dapps can query rewards period status on chain.